File: lib/safe-withdrawal.ts
import {
  calculateCompoundInterest,
  type CompoundContext,
  type CompoundResult,
} from "./compound-interest";
import type { CompoundParams } from "./portfolio-types";

export interface SafeWithdrawalAdvice {
  /** Макс. % / год от номинальной ликвидности */
  maxAnnualPercent: number;
  /** Макс. ₽ / мес в сегодняшних рублях (фикс. режим) */
  maxMonthlyReal: number;
  /** Ликвидная часть в первый месяц вывода, номинал (до списания) */
  liquidityAtWithdrawalStart: number;
  /** Та же ликвидность в сегодняшних ₽ */
  liquidityAtWithdrawalStartReal: number;
  /** Текущий сценарий укладывается в безопасный диапазон */
  currentIsSafe: boolean;
  /**
   * Эквивалент maxMonthlyReal в % / год от номинальной ликвидности
   * (целевой номинал = сумма в сегодняшних ₽ × инфляция на старт вывода)
   */
  maxMonthlyAsNominalPercent: number;
  /** Первая выплата при maxAnnualPercent, сегодняшние ₽ (на руки) */
  maxPercentAsMonthlyReal: number;
  /** Последняя выплата при maxAnnualPercent, сегодняшние ₽ (на руки) */
  maxPercentAsMonthlyRealEnd: number;
  /** Первая выплата при текущих параметрах, сегодняшние ₽ */
  currentStartPayoutReal: number;
  /** Эквивалент текущей фикс. суммы в % от номинала (если режим fixed) */
  currentFixedAsNominalPercent: number | null;
  /** Эквивалент текущего % в ₽/мес сегодня (если режим percent) */
  currentPercentAsMonthlyReal: number | null;
}

function monthlyInflationRate(params: CompoundParams): number {
  const annual = params.inflationPercent / 100;
  if ((params.monthlyRateMethod ?? "effective") === "simple") {
    return annual / 12;
  }
  return (1 + annual) ** (1 / 12) - 1;
}

function firstWithdrawalMonth(params: CompoundParams): number {
  return Math.round((params.withdrawAfterYears ?? 0) * 12) + 1;
}

function inflationFactorAtWithdrawalStart(params: CompoundParams): number {
  const month = firstWithdrawalMonth(params);
  return (1 + monthlyInflationRate(params)) ** month;
}

/** % / год от номинальной ликвидности для фикс. суммы в сегодняшних ₽ (до налога) */
export function fixedRealToNominalPercent(
  monthlyReal: number,
  nominalLiquidity: number,
  params: CompoundParams,
): number {
  if (nominalLiquidity <= 0 || monthlyReal <= 0) return 0;
  const nominalMonthlyTarget = monthlyReal * inflationFactorAtWithdrawalStart(params);
  return (nominalMonthlyTarget * 12 * 100) / nominalLiquidity;
}

function realLiquidityAtPoint(
  month: number,
  liquidityBalance: number,
  params: CompoundParams,
): number {
  const factor = (1 + monthlyInflationRate(params)) ** month;
  return liquidityBalance / factor;
}

/** Ликвидность не обнуляется и не падает в сегодняшних ₽ за фазу вывода */
export function isWithdrawalSustainable(
  result: CompoundResult,
  params: CompoundParams,
): boolean {
  if (result.withdrawalEndedEarly || result.withdrawalMonthsLiquidityEmpty > 0) {
    return false;
  }

  const withdrawalPoints = result.points.filter((p) => p.inWithdrawalPhase);
  if (withdrawalPoints.length < 2) {
    return true;
  }

  const first = withdrawalPoints[0];
  const last = withdrawalPoints[withdrawalPoints.length - 1];
  const firstReal = realLiquidityAtPoint(
    first.month,
    first.liquidityBalance,
    params,
  );
  const lastReal = realLiquidityAtPoint(
    last.month,
    last.liquidityBalance,
    params,
  );

  return lastReal >= firstReal * 0.995;
}

function getBaselineAtWithdrawalStart(
  params: CompoundParams,
  context?: CompoundContext,
): CompoundResult {
  return calculateCompoundInterest(
    {
      ...params,
      withdrawalMode: "percent",
      annualWithdrawalPercent: 0,
      monthlyWithdrawal: 0,
    },
    context,
  );
}

function getLiquidityAtWithdrawalStart(
  params: CompoundParams,
  context?: CompoundContext,
): number {
  return getBaselineAtWithdrawalStart(params, context).withdrawalStartLiquidity ?? 0;
}

function simulatePercent(
  params: CompoundParams,
  annualPercent: number,
  context?: CompoundContext,
): CompoundResult {
  return calculateCompoundInterest(
    {
      ...params,
      withdrawalMode: "percent",
      annualWithdrawalPercent: annualPercent,
      monthlyWithdrawal: 0,
    },
    context,
  );
}

function simulateFixed(
  params: CompoundParams,
  monthlyReal: number,
  context?: CompoundContext,
): CompoundResult {
  return calculateCompoundInterest(
    {
      ...params,
      withdrawalMode: "fixed",
      monthlyWithdrawal: monthlyReal,
      annualWithdrawalPercent: 0,
    },
    context,
  );
}

function findMaxAnnualPercent(
  params: CompoundParams,
  context?: CompoundContext,
): number {
  let lo = 0;
  let hi = Math.max(params.annualReturnPercent, 1);

  while (
    hi < 100 &&
    isWithdrawalSustainable(simulatePercent(params, hi, context), params)
  ) {
    lo = hi;
    hi = Math.min(100, hi * 2);
  }

  for (let i = 0; i < 40; i++) {
    const mid = (lo + hi) / 2;
    if (isWithdrawalSustainable(simulatePercent(params, mid, context), params)) {
      lo = mid;
    } else {
      hi = mid;
    }
    if (hi - lo < 0.01) break;
  }

  return Math.round(lo * 100) / 100;
}

function findMaxMonthlyReal(
  params: CompoundParams,
  context?: CompoundContext,
  liquidityAtStart?: number,
): number {
  const liquidity =
    liquidityAtStart ?? getLiquidityAtWithdrawalStart(params, context);
  const yieldBased =
    (liquidity * (params.annualReturnPercent / 100)) / 12;
  let lo = 0;
  let hi = Math.max(yieldBased * 2, params.monthlyWithdrawal || 0, 10_000);

  while (
    hi < 50_000_000 &&
    isWithdrawalSustainable(simulateFixed(params, hi, context), params)
  ) {
    lo = hi;
    hi *= 2;
  }

  for (let i = 0; i < 40; i++) {
    const mid = (lo + hi) / 2;
    if (isWithdrawalSustainable(simulateFixed(params, mid, context), params)) {
      lo = mid;
    } else {
      hi = mid;
    }
    if (hi - lo < 1) break;
  }

  return Math.floor(lo);
}

function currentScenarioIsSafe(
  params: CompoundParams,
  context: CompoundContext | undefined,
  maxAnnualPercent: number,
  maxMonthlyReal: number,
): boolean {
  const mode = params.withdrawalMode ?? "fixed";
  if (mode === "percent") {
    const pct = params.annualWithdrawalPercent ?? 0;
    if (pct <= 0) return true;
    return (
      pct <= maxAnnualPercent + 0.01 &&
      isWithdrawalSustainable(
        simulatePercent(params, pct, context),
        params,
      )
    );
  }

  const monthly = params.monthlyWithdrawal ?? 0;
  if (monthly <= 0) return true;
  return (
    monthly <= maxMonthlyReal + 1 &&
    isWithdrawalSustainable(simulateFixed(params, monthly, context), params)
  );
}

export function computeSafeWithdrawalAdvice(
  params: CompoundParams,
  context?: CompoundContext,
): SafeWithdrawalAdvice | null {
  if (params.withdrawAfterYears == null || params.withdrawAfterYears <= 0) {
    return null;
  }

  const inflationAtStart = inflationFactorAtWithdrawalStart(params);
  const liquidityAtWithdrawalStart = getLiquidityAtWithdrawalStart(
    params,
    context,
  );
  const liquidityAtWithdrawalStartReal =
    liquidityAtWithdrawalStart / inflationAtStart;
  const maxAnnualPercent = findMaxAnnualPercent(params, context);
  const maxMonthlyReal = findMaxMonthlyReal(
    params,
    context,
    liquidityAtWithdrawalStart,
  );

  const maxMonthlyAsNominalPercent = fixedRealToNominalPercent(
    maxMonthlyReal,
    liquidityAtWithdrawalStart,
    params,
  );
  const maxPercentSim = simulatePercent(params, maxAnnualPercent, context);
  const maxPercentAsMonthlyReal = maxPercentSim.withdrawalStartPayoutReal;
  const maxPercentAsMonthlyRealEnd = maxPercentSim.withdrawalPayoutReal;

  const mode = params.withdrawalMode ?? "fixed";
  const currentSim =
    mode === "percent"
      ? simulatePercent(params, params.annualWithdrawalPercent ?? 0, context)
      : simulateFixed(params, params.monthlyWithdrawal ?? 0, context);

  const currentStartPayoutReal = currentSim.withdrawalStartPayoutReal;
  const currentFixedAsNominalPercent =
    mode === "fixed" && (params.monthlyWithdrawal ?? 0) > 0
      ? fixedRealToNominalPercent(
          params.monthlyWithdrawal ?? 0,
          liquidityAtWithdrawalStart,
          params,
        )
      : null;
  const currentPercentAsMonthlyReal =
    mode === "percent" ? currentStartPayoutReal : null;

  return {
    maxAnnualPercent,
    maxMonthlyReal,
    liquidityAtWithdrawalStart,
    liquidityAtWithdrawalStartReal,
    currentIsSafe: currentScenarioIsSafe(
      params,
      context,
      maxAnnualPercent,
      maxMonthlyReal,
    ),
    maxMonthlyAsNominalPercent: Math.round(maxMonthlyAsNominalPercent * 100) / 100,
    maxPercentAsMonthlyReal,
    maxPercentAsMonthlyRealEnd,
    currentStartPayoutReal,
    currentFixedAsNominalPercent:
      currentFixedAsNominalPercent != null
        ? Math.round(currentFixedAsNominalPercent * 100) / 100
        : null,
    currentPercentAsMonthlyReal,
  };
}

File: crates/finance-core/src/compound/safe_withdrawal.rs
//! Safe-withdrawal search mirroring `lib/safe-withdrawal.ts`.
#![allow(clippy::manual_midpoint)]

use serde::Serialize;

use super::{
    rates::monthly_rate_from_annual,
    simulate::{calculate_compound_interest, CompoundError},
    types::{CompoundContext, CompoundOptions, CompoundParams, CompoundResult, WithdrawalMode},
};

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SafeWithdrawalAdvice {
    pub max_annual_percent: f64,
    pub max_monthly_real: f64,
    pub liquidity_at_withdrawal_start: f64,
    pub liquidity_at_withdrawal_start_real: f64,
    pub current_is_safe: bool,
    pub max_monthly_as_nominal_percent: f64,
    pub max_percent_as_monthly_real: f64,
    pub max_percent_as_monthly_real_end: f64,
    pub current_start_payout_real: f64,
    pub current_fixed_as_nominal_percent: Option<f64>,
    pub current_percent_as_monthly_real: Option<f64>,
}

fn monthly_inflation_rate(params: &CompoundParams) -> f64 {
    monthly_rate_from_annual(params.inflation_percent, params.monthly_rate_method)
}

fn first_withdrawal_month(params: &CompoundParams) -> u32 {
    let years = params.withdraw_after_years.unwrap_or(0.0);
    u32::try_from((years * 12.0).round() as i64 + 1).unwrap_or(u32::MAX)
}

fn inflation_factor_at_withdrawal_start(params: &CompoundParams) -> f64 {
    (1.0 + monthly_inflation_rate(params))
        .powi(i32::try_from(first_withdrawal_month(params)).unwrap_or(i32::MAX))
}

pub fn fixed_real_to_nominal_percent(
    monthly_real: f64,
    nominal_liquidity: f64,
    params: &CompoundParams,
) -> f64 {
    if nominal_liquidity <= 0.0 || monthly_real <= 0.0 {
        return 0.0;
    }
    let nominal_monthly_target = monthly_real * inflation_factor_at_withdrawal_start(params);
    (nominal_monthly_target * 12.0 * 100.0) / nominal_liquidity
}

fn real_liquidity_at_point(month: u32, liquidity_balance: f64, params: &CompoundParams) -> f64 {
    let factor =
        (1.0 + monthly_inflation_rate(params)).powi(i32::try_from(month).unwrap_or(i32::MAX));
    liquidity_balance / factor
}

pub fn is_withdrawal_sustainable(result: &CompoundResult, params: &CompoundParams) -> bool {
    if result.withdrawal_ended_early || result.withdrawal_months_liquidity_empty > 0 {
        return false;
    }

    let withdrawal_points: Vec<_> = result
        .points
        .iter()
        .filter(|point| point.in_withdrawal_phase)
        .collect();
    if withdrawal_points.len() < 2 {
        return true;
    }

    let first = withdrawal_points[0];
    let last = withdrawal_points[withdrawal_points.len() - 1];
    let first_real = real_liquidity_at_point(first.month, first.liquidity_balance, params);
    let last_real = real_liquidity_at_point(last.month, last.liquidity_balance, params);
    last_real >= first_real * 0.995
}

fn simulate_percent(
    params: &CompoundParams,
    annual_percent: f64,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
) -> Result<CompoundResult, CompoundError> {
    let mut next = params.clone();
    next.withdrawal_mode = WithdrawalMode::Percent;
    next.annual_withdrawal_percent = annual_percent;
    next.monthly_withdrawal = 0.0;
    calculate_compound_interest(&next, context, options)
}

fn simulate_fixed(
    params: &CompoundParams,
    monthly_real: f64,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
) -> Result<CompoundResult, CompoundError> {
    let mut next = params.clone();
    next.withdrawal_mode = WithdrawalMode::Fixed;
    next.monthly_withdrawal = monthly_real;
    next.annual_withdrawal_percent = 0.0;
    calculate_compound_interest(&next, context, options)
}

fn get_liquidity_at_withdrawal_start(
    params: &CompoundParams,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
) -> Result<f64, CompoundError> {
    Ok(simulate_percent(params, 0.0, context, options)?
        .withdrawal_start_liquidity
        .unwrap_or(0.0))
}

fn find_max_annual_percent(
    params: &CompoundParams,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
) -> Result<f64, CompoundError> {
    let mut lo = 0.0;
    let mut hi = params.annual_return_percent.max(1.0);

    while hi < 100.0
        && is_withdrawal_sustainable(&simulate_percent(params, hi, context, options)?, params)
    {
        lo = hi;
        hi = (hi * 2.0).min(100.0);
    }

    for _ in 0..40 {
        let mid = (lo + hi) / 2.0;
        if is_withdrawal_sustainable(&simulate_percent(params, mid, context, options)?, params) {
            lo = mid;
        } else {
            hi = mid;
        }
        if hi - lo < 0.01 {
            break;
        }
    }

    Ok((lo * 100.0).round() / 100.0)
}

fn find_max_monthly_real(
    params: &CompoundParams,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
    liquidity_at_start: f64,
) -> Result<f64, CompoundError> {
    let yield_based = (liquidity_at_start * (params.annual_return_percent / 100.0)) / 12.0;
    let mut lo = 0.0;
    let mut hi = yield_based * 2.0;
    if params.monthly_withdrawal > hi {
        hi = params.monthly_withdrawal;
    }
    if hi < 10_000.0 {
        hi = 10_000.0;
    }

    while hi < 50_000_000.0
        && is_withdrawal_sustainable(&simulate_fixed(params, hi, context, options)?, params)
    {
        lo = hi;
        hi *= 2.0;
    }

    for _ in 0..40 {
        let mid = (lo + hi) / 2.0;
        if is_withdrawal_sustainable(&simulate_fixed(params, mid, context, options)?, params) {
            lo = mid;
        } else {
            hi = mid;
        }
        if hi - lo < 1.0 {
            break;
        }
    }

    Ok(lo.floor())
}

fn current_scenario_is_safe(
    params: &CompoundParams,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
    max_annual_percent: f64,
    max_monthly_real: f64,
) -> Result<bool, CompoundError> {
    match params.withdrawal_mode {
        WithdrawalMode::Percent => {
            let pct = params.annual_withdrawal_percent;
            if pct <= 0.0 {
                return Ok(true);
            }
            Ok(pct <= max_annual_percent + 0.01
                && is_withdrawal_sustainable(
                    &simulate_percent(params, pct, context, options)?,
                    params,
                ))
        }
        WithdrawalMode::Fixed => {
            let monthly = params.monthly_withdrawal;
            if monthly <= 0.0 {
                return Ok(true);
            }
            Ok(monthly <= max_monthly_real + 1.0
                && is_withdrawal_sustainable(
                    &simulate_fixed(params, monthly, context, options)?,
                    params,
                ))
        }
    }
}

pub fn compute_safe_withdrawal_advice(
    params: &CompoundParams,
    context: Option<&CompoundContext>,
    options: &CompoundOptions,
) -> Result<Option<SafeWithdrawalAdvice>, CompoundError> {
    let Some(years) = params.withdraw_after_years else {
        return Ok(None);
    };
    if years <= 0.0 {
        return Ok(None);
    }

    let inflation_at_start = inflation_factor_at_withdrawal_start(params);
    let liquidity_at_withdrawal_start =
        get_liquidity_at_withdrawal_start(params, context, options)?;
    let liquidity_at_withdrawal_start_real = liquidity_at_withdrawal_start / inflation_at_start;
    let max_annual_percent = find_max_annual_percent(params, context, options)?;
    let max_monthly_real =
        find_max_monthly_real(params, context, options, liquidity_at_withdrawal_start)?;
    let max_monthly_as_nominal_percent =
        fixed_real_to_nominal_percent(max_monthly_real, liquidity_at_withdrawal_start, params);
    let max_percent_sim = simulate_percent(params, max_annual_percent, context, options)?;
    let current_sim = match params.withdrawal_mode {
        WithdrawalMode::Percent => {
            simulate_percent(params, params.annual_withdrawal_percent, context, options)?
        }
        WithdrawalMode::Fixed => {
            simulate_fixed(params, params.monthly_withdrawal, context, options)?
        }
    };
    let current_start_payout_real = current_sim.withdrawal_start_payout_real;
    let current_fixed_as_nominal_percent =
        if params.withdrawal_mode == WithdrawalMode::Fixed && params.monthly_withdrawal > 0.0 {
            Some(
                (fixed_real_to_nominal_percent(
                    params.monthly_withdrawal,
                    liquidity_at_withdrawal_start,
                    params,
                ) * 100.0)
                    .round()
                    / 100.0,
            )
        } else {
            None
        };
    let current_percent_as_monthly_real = if params.withdrawal_mode == WithdrawalMode::Percent {
        Some(current_start_payout_real)
    } else {
        None
    };

    Ok(Some(SafeWithdrawalAdvice {
        max_annual_percent,
        max_monthly_real,
        liquidity_at_withdrawal_start,
        liquidity_at_withdrawal_start_real,
        current_is_safe: current_scenario_is_safe(
            params,
            context,
            options,
            max_annual_percent,
            max_monthly_real,
        )?,
        max_monthly_as_nominal_percent: (max_monthly_as_nominal_percent * 100.0).round() / 100.0,
        max_percent_as_monthly_real: max_percent_sim.withdrawal_start_payout_real,
        max_percent_as_monthly_real_end: max_percent_sim.withdrawal_payout_real,
        current_start_payout_real,
        current_fixed_as_nominal_percent,
        current_percent_as_monthly_real,
    }))
}
